### DISCUSSION TITLE
[分享] 关于 `TypeError: o is not a constructor`

### DISCUSSION THREAD

**Discussion by Tinhone** [the1812/Bilibili-Evolved#4877] [分享 / Share]
这个问题是因自动更新了组件`简化评论区`但没有更新脚本本体引起的。

[这次更新中的新代码](https://github.com/the1812/Bilibili-Evolved/blob/334026be9b6ef221965066a3fad472b58b6a483b/registry/lib/components/style/simplify/comments/index.ts#L56)
```javascript
const { ShadowDomStyles } = await import('@/core/shadow-dom')
const v3Styles = await import('./comments-v3.scss').then(m => m.default)
const shadowDom = new ShadowDomStyles()
shadowDom.addStyle(v3Styles)
```

编译产物中对应的代码
```javascript
const {
  ShadowDomStyles: o
}
= await Promise.resolve().then(i.t.bind(i, 488, 23)),
m = await Promise.resolve().then(i.t.bind(i, 62, 23)).then((n=>n.default));
(new o).addStyle(m)
```

因为 `core/shadow-dom` API 在旧版本脚本本体中不存在（`o === undefined`），所以显然会报错 `TypeError: o is not a constructor`。

不知道怎么回事报错堆栈定位不到关键代码，挺奇怪的。

<details>
<summary>报错堆栈</summary>
<pre>
Bilibili Evolved TypeError: o is not a constructor
    entry 404 line 9 > injectedScript line 197 > Function line 4 > eval:9
    entry Bilibili-Evolved-Preview.user.js:11886
    entry Bilibili-Evolved-Preview.user.js:12987
    a Bilibili-Evolved-Preview.user.js:6174
    h Bilibili-Evolved-Preview.user.js:6204
    a Bilibili-Evolved-Preview.user.js:6243
    promise callback*a Bilibili-Evolved-Preview.user.js:6243
    F Bilibili-Evolved-Preview.user.js:6259
    r Bilibili-Evolved-Preview.user.js:12608
    r Bilibili-Evolved-Preview.user.js:12608
    r Bilibili-Evolved-Preview.user.js:12607
    F Bilibili-Evolved-Preview.user.js:6259
    F Bilibili-Evolved-Preview.user.js:6258
    async*window["__f__lzwc5452.ll"]/</</</</< Bilibili-Evolved-Preview.user.js:36693
    r Bilibili-Evolved-Preview.user.js:13043
    async*window["__f__lzwc5452.ll"]/</</</< Bilibili-Evolved-Preview.user.js:36688
    async*window["__f__lzwc5452.ll"]/</</< Bilibili-Evolved-Preview.user.js:36729
    window["__f__lzwc5452.ll"]/</< Bilibili-Evolved-Preview.user.js:36730
    window["__f__lzwc5452.ll"]/< Bilibili-Evolved-Preview.user.js:36731
    wt 404:9
    window["__f__lzwc5452.ll"]/< Bilibili-Evolved-Preview.user.js:9
    "__f__lzwc5452.ll" Bilibili-Evolved-Preview.user.js:31
    wt 404:9
    o 404:81
    <anonymous> 404:83
    set 404:91
    <anonymous> Bilibili-Evolved-Preview.user.js:1
    wt 404:9
    f 404:77
    l 404:79
    send 404:80
    Wn 404:17
    send 404:79
    <anonymous> 404:83
Bilibili-Evolved-Preview.user.js:16751
    b Bilibili-Evolved-Preview.user.js:16751
    window["__f__lzwc5452.ll"]/</</</</</< Bilibili-Evolved-Preview.user.js:250
    (异步：promise callback)
    window["__f__lzwc5452.ll"]/</</</</< Bilibili-Evolved-Preview.user.js:250
    r Bilibili-Evolved-Preview.user.js:13043
    window["__f__lzwc5452.ll"]/</</</< Bilibili-Evolved-Preview.user.js:36688
    window["__f__lzwc5452.ll"]/</</< Bilibili-Evolved-Preview.user.js:36729
    window["__f__lzwc5452.ll"]/</< Bilibili-Evolved-Preview.user.js:36730
    window["__f__lzwc5452.ll"]/< Bilibili-Evolved-Preview.user.js:36731
    wt 404:9
    window["__f__lzwc5452.ll"]/< Bilibili-Evolved-Preview.user.js:1
    "__f__lzwc5452.ll" Bilibili-Evolved-Preview.user.js:1
    wt 404:9
    o 404:81
    <anonymous> 404:83
    set 404:91
    <anonymous> Bilibili-Evolved-Preview.user.js:1
    wt 404:9
    f 404:77
    l 404:79
    send 404:80
    Wn 404:17
    send 404:79
    <anonymous> 404:83
</pre>
</details>

~~_还好不是浏览器兼容性问题_~~

**Comment by the1812** [comment]
真正的问题应该在于自动更新有 bug, 组件版本大于本体时不应该更新
目前所有产物都会注明要求的版本, 来避免这种情况, 但是确实遇到过有些人仍然装到了高版本组件的问题
![Snipaste_2024-08-16_17-36-13](https://github.com/user-attachments/assets/0f8a95c6-6ca7-44bb-a223-1d16ebfe22f4)

**Reply by Tinhone** [reply]
~~_感觉可以在组件编译产物或者 `defineComponentMetadata` 里加个默认为 true 的说明是否是破坏性变更的键，但不知道好不好做。这样的话在组件更新没涉及本体的时候，手动写或者自动加一下就可以提早更新了 🤔_~~

对了，上面提到的异常看起来还会阻断 [`components block time` 和 `components resolve time`](https://github.com/the1812/Bilibili-Evolved/blob/334026be9b6ef221965066a3fad472b58b6a483b/src/components/component.ts#L180) 的日志输出，也许可以在哪里加一个错误处理？

**Reply by Tinhone** [reply]
原来有时候会安装高版本组件不是预期情况啊 ( ﾟ∀。)

**Reply by the1812** [reply]
调试了下这个检测逻辑也没发现有啥问题, 就加了个 log, 更新时可以留意一下
![Snipaste_2024-09-08_23-42-39](https://github.com/user-attachments/assets/0f634a80-fadb-4a42-8875-927a7c3d5d13)
